utils/cache.js

```diff
@@ -0,0 +1,113 @@
+
+/**
+ * Cache pool to get and store API responses
+ * @param {Object} [opt] options for cache pool
+ * @param {number} opt.maximumSize Maximum size of the cache pool in Megabytes (MB)
+ */
+const cachePool = (opt = { maximumSize: CACHESIZE }) => {
+  const cacheStore = new Map();
+  let hits = 0;
+
+  /**
+   * Get an API Respnose from cacheStore.
+   * @param {string} key
+   * @returns {null | object}
+   */
+  const get = (key) => {
+    const apiData = cacheStore.get(key);
+
+    if (!apiData) {
+      return null;
+    }
+
+    const isApiDataExpired = new Date().getTime() > apiData.expiry;
+
+    // If data is expired remove it from store, time to get a fresh copy.
+    if (isApiDataExpired) {
+      cacheStore.delete(key);
+      return null;
+    }
+
+    hits += 1;
+    return apiData.response;
+  };
+
+  /**
+   * Add API response to cacheStore.
+   * @param {string} key
+   * @param {Object} value Value to be stored inside the cache.
+   * @param {number} value.priority Priority of the api
+   * @param {number} value.expiry Expiry time of api
+   * @param {number} value.size Size of api response in byte
+   * @returns {number} : statusCode
+   */
+  const set = async (key, value) => {
+    try {
+      cacheStore.set(key, value);
+      return 200;
+    } catch (err) {
+      return 500;
+    }
+  };
+
+  return { get, set, hits, cacheStore };
+};
+
+// Initialize cache pool.
+const pool = cachePool();
+
+/**
+ * Caching middleware for API resposnes.
+ * @param {Object} [data] Options to handle individual api cache.
+ * @param {number} data.priority The priority of api in cache store.
+ * @param {number} data.expiry Cache expiry time of api in minutes.
+ * @returns {function} middleware function to help cache api response.
+ */
+const cache = (data = { priority: 2, expiry: CACHEEXPIRYTIME }) => {
+  return async (req, res, next) => {
+    const key = "__cache__" + req.method + req.originalUrl;
+    const dataInCache = pool.get(key);
+
+    if (dataInCache) {
+      res.send(dataInCache);
+    } else {
+      /**
+       * As we do not have data in our cache we call the next middleware,
+       * intercept the response being sent from middleware and store it in cache.
+       *  */
+      const chunks = [];
+      const oldWrite = res.write;
+      const oldEnd = res.end;
+
+      res.write = (chunk, ...args) => {
+        chunks.push(chunk);
+        return oldWrite.apply(res, [chunk, ...args]);
+      };
+
+      res.end = (chunk, ...args) => {
+        if (chunk) {
+          chunks.push(chunk);
+        }
+
+        const apiResponse = Buffer.concat(chunks).toString();
+
+        const cacheValue = {
+          priority: data.priority,
+          response: apiResponse,
+          expiry: new Date().getTime() + minutesToMilliseconds(data.expiry),
+          size: Buffer.byteLength(apiResponse),
+        };
+
+        pool.set(key, cacheValue);
+        return oldEnd.apply(res, [chunk, ...args]);
+      };
```

Comment:
Instead of intercepting request can we have a middleware after the request handler in the route that'll check if we need to write the value in cache or not?

Reply:
Here we are Intercepting request to get the response being sent from the controller of that route and then store its response in the cache, not sure how would it be done using middleware after the request handler. 

Would you please give an example of your implementation?

Reply:
I guess we’ll have to use this intercepting way of capturing the response body. I couldn’t find a better way of doing this

Reply:
Isn't Middleware is called only before the route handler is called. And to have access to the response object, but  don't have the result of the route handler. So i think the intercepting way is way more good.